src/decode.rs

```diff
@@ -454,6 +456,226 @@ impl<T: Read, O: Read> fmt::Debug for DecoderShared<T, O> {
+                    } => {
+                        // todo: add direct output optimization
+
+                        // ensure reading state, reading size bytes
+                        // we might already be in the reading state,
+                        // so we must not set buf_start to 0
+                        self.buf_end = size;
+                        // chunk never comes from outboard
+                        ready!(self.poll_fill_buffer_from_input(cx))?;
+
+                        // Hash it and push its hash into the VerifyState. This
+                        // returns an error if the hash is bad. Otherwise, the
+                        // chunk is verified.
+                        let read_buf = &self.buf[0..size];
+                        let chunk_hash = blake3::guts::ChunkState::new(index)
+                            .update(read_buf)
+                            .finalize(finalization.is_root());
+                        self.state.feed_chunk(&chunk_hash)?;
+
+                        // we go into the writing state now, starting from skip
+                        self.buf_start = skip;
+                        // we should have something to write,
+                        // unless the entire chunk was empty
+                        debug_assert!(self.buf_len() > 0 || size == 0);
+                    }
+                }
+            }
+        }
+
+        fn poll_fill_buffer_from_input(
+            &mut self,
+            cx: &mut task::Context<'_>,
+        ) -> task::Poll<Result<(), io::Error>> {
+            let mut buf = ReadBuf::new(&mut self.buf[..self.buf_end]);
+            buf.advance(self.buf_start);
+            let src = &mut self.input;
+            while buf.remaining() > 0 {
+                ready!(AsyncRead::poll_read(Pin::new(src), cx, &mut buf))?;
+                self.buf_start = buf.filled().len();
+            }
+            task::Poll::Ready(Ok(()))
+        }
+
+        fn poll_fill_buffer_from_outboard(
+            &mut self,
+            cx: &mut task::Context<'_>,
+        ) -> task::Poll<Result<(), io::Error>> {
+            let mut buf = ReadBuf::new(&mut self.buf[..self.buf_end]);
+            buf.advance(self.buf_start);
+            let src = self.outboard.as_mut().unwrap();
+            while buf.remaining() > 0 {
+                ready!(AsyncRead::poll_read(Pin::new(src), cx, &mut buf))?;
+                self.buf_start = buf.filled().len();
+            }
+            task::Poll::Ready(Ok(()))
+        }
+
+        fn poll_fill_buffer_from_input_or_outboard(
+            &mut self,
+            cx: &mut task::Context<'_>,
+        ) -> task::Poll<Result<(), io::Error>> {
+            if self.outboard.is_some() {
+                self.poll_fill_buffer_from_outboard(cx)
+            } else {
+                self.poll_fill_buffer_from_input(cx)
+            }
+        }
+    }
+
+    #[derive(Clone, Debug)]
+    pub struct AsyncDecoder<T: AsyncRead + Unpin, O: AsyncRead + Unpin> {
```

Comment:
It would also be possible to have the Decoder support both async and sync, but the ergonomics of that is really bad. I tried it. So I went with this tiny wrapper instead.